File: backend/internal/audit/audit.go
// Package audit 实现《统一审计留痕（audit log）跨服务契约》的写入侧。
//
// catalog / auth / community / storage 四仓各复制一份**同源代码**：四仓是独立 module，没有跨仓
// 依赖通道，所以这里的 ServiceName / Schema / 函数名与语义必须与其他三份保持一致，改动要四仓同步
// （契约 §3）。本包只依赖标准库、gin 与 uuid，不引用任何服务自己的类型，才能被逐字复制过去。
package audit

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// ServiceName 是本服务写进 audit_log.service 的值（契约 §1：该列无 CHECK，新增服务不改旧 DDL）。
const ServiceName = "catalog"

// Schema 是契约 §1 的 DDL 逐字复制：CREATE SCHEMA / advisory 锁（740205，四个服务共用，
// 用来串行化"四个服务同时首次建表"）/ 建表 / 建索引，全部 IF NOT EXISTS，重复执行安全。
const Schema = `-- 审计表跨服务共用：四个服务的业务 DDL 各管自己的 schema，这里单独用 audit schema，
-- 因为它不属于任何单个服务的领域数据（见 §5 的取舍说明）。
CREATE SCHEMA IF NOT EXISTS audit;

-- 四个服务可能同时首次启动；建表用同一个 advisory 锁键（740205）串行化。
-- 一次 Exec 里的多条语句由 lib/pq 作为隐式事务批处理发送，xact 锁因此覆盖到建表结束。
SELECT pg_advisory_xact_lock(740205);

CREATE TABLE IF NOT EXISTS audit.audit_log (
  id               uuid PRIMARY KEY,
  occurred_at      timestamptz NOT NULL DEFAULT now(),
  service          text NOT NULL,
  action           text NOT NULL,
  actor_user_id    uuid,
  actor_username   text NOT NULL DEFAULT '',
  credential_type  text NOT NULL DEFAULT '',
  actor_ip         text NOT NULL DEFAULT '',
  actor_user_agent text NOT NULL DEFAULT '',
  target_type      text NOT NULL DEFAULT '',
  target_id        text NOT NULL DEFAULT '',
  changes          jsonb NOT NULL DEFAULT '{}'::jsonb,
  result           text NOT NULL DEFAULT 'success' CHECK (result IN ('success','failure')),
  error_code       text NOT NULL DEFAULT '',
  request_method   text NOT NULL DEFAULT '',
  route            text NOT NULL DEFAULT '',
  http_status      int NOT NULL DEFAULT 0,
  request_id       text NOT NULL DEFAULT ''
);
CREATE INDEX IF NOT EXISTS audit_log_occurred_at_idx ON audit.audit_log(occurred_at DESC);
CREATE INDEX IF NOT EXISTS audit_log_service_action_idx ON audit.audit_log(service, action, occurred_at DESC);
CREATE INDEX IF NOT EXISTS audit_log_actor_idx ON audit.audit_log(actor_user_id, occurred_at DESC);
CREATE INDEX IF NOT EXISTS audit_log_target_idx ON audit.audit_log(target_type, target_id, occurred_at DESC);
`

// 契约写死的上限与上下文键（§1、§3、§4）。
const (
	// QueueSize 是有界队列容量：满则丢行并计数，绝不阻塞业务响应。
	QueueSize = 1024
	// ErrorCodeKey 是失败错误码的 gin 上下文键：处理器用 Fail 写入，中间件在 c.Next() 之后读它。
	ErrorCodeKey = "audit_error_code"
	// RequestIDHeader 是请求 id 的透传头：请求没带时由中间件生成 uuid 并回写同名响应头。
	RequestIDHeader = "X-Request-Id"

	userAgentMax = 512
	valueMax     = 512
	changesMax   = 8 * 1024
	// insertTimeout 是单行落库的上限：后台 goroutine 不能被一次卡死的写入拖住整条队列。
	insertTimeout = 5 * time.Second
)

const (
	redacted     = "[redacted]"
	truncatedKey = "_truncated"
	// detailKey 是 Describe 与中间件之间的上下文键。
	detailKey = "audit_detail"
)

// Entry 是一行审计（字段与契约 §1 的列一一对应）。ID / OccurredAt / Service / Result 为空时
// 由 Recorder 补齐：service 取 Recorder 的服务名，occurred_at 取当前时间，result 取 success。
type Entry struct {
	ID             string
	OccurredAt     time.Time
	Service        string
	Action         string
	ActorUserID    string
	ActorUsername  string
	CredentialType string
	ActorIP        string
	ActorUserAgent string
	TargetType     string
	TargetID       string
	Changes        map[string]any
	Result         string
	ErrorCode      string
	RequestMethod  string
	Route          string
	HTTPStatus     int
	RequestID      string
}

// Detail 是处理器补充的被动对象与变更摘要（契约 §3）：中间件按响应状态定 result，
// 这里的 target/changes 原样进审计行。
type Detail struct {
	TargetType string
	TargetID   string
	Changes    map[string]any
}

// Actor 是操作者快照：UserID 为空表示匿名。CredentialType 取 session / pat / oauth /
// anonymous / system；非账号服务只能给 pat 或 session 这类近似值（契约 §7）。
type Actor struct {
	UserID         string
	Username       string
	CredentialType string
}

// Options 是中间件的接线参数。
type Options struct {
	Recorder *Recorder
	// Actions 是"METHOD 路由模板" → 动作码。只有登记在册的请求才写审计（GET 与豁免路由不写）。
	Actions map[string]string
	// Exempt 是"METHOD 路由模板" → 豁免理由：用了写方法但没有写入语义的路由。
	// 运行期不读它（未登记的路由本就不写），它由各服务的「写路由覆盖守卫测试」读取，
	// 让"这条写路由为什么不记"有据可查，而不是被忘掉。
	Exempt map[string]string
	// Actor 从请求上下文取操作者；为 nil 时操作者留空（等同匿名）。
	Actor func(*gin.Context) Actor
}

// Recorder 是审计写入器：一个后台 goroutine + 有界 channel。
// 零值不可用，必须经 NewRecorder 构造。
type Recorder struct {
	db      *sql.DB
	service string
	queue   chan Entry
	wg      sync.WaitGroup
	dropped atomic.Int64
	// mu + closed 把"Record 入队"与"Close 关通道"串行化：没有它，Close 之后到达的 Record
	// 会往已关闭的 channel 发送而 panic（业务侧不该因为收尾而崩）。
	mu     sync.RWMutex
	closed bool
}

// NewRecorder 起后台写库 goroutine；db 可为 nil（单测未接库），此时落库会返回错误并被丢弃。
func NewRecorder(db *sql.DB, service string) *Recorder {
	r := &Recorder{db: db, service: service, queue: make(chan Entry, QueueSize)}
	r.wg.Add(1)
	go r.loop()
	return r
}

// Record 非阻塞入队：队列满时打 error 日志并丢弃，绝不阻塞业务响应（契约 §3）。
func (r *Recorder) Record(e Entry) {
	if r == nil {
		return
	}
	r.mu.RLock()
	defer r.mu.RUnlock()
	if r.closed {
		// 进程收尾阶段（Close 之后）到达的行直接丢：此时没有消费者了，入队只会堆积。
		return
	}
	select {
	case r.queue <- e:
	default:
		n := r.dropped.Add(1)
		slog.Error("审计入队失败：队列已满，丢弃该行（业务不受影响）",
			"service", r.service, "action", e.Action, "target_type", e.TargetType, "dropped_total", n)
	}
}

// RecordSync 同步写一行（只给测试与"必须强一致"的少数动作用；本服务没有强一致动作）。
func (r *Recorder) RecordSync(ctx context.Context, e Entry) error {
	if r == nil {
		return errors.New("audit: recorder unavailable")
	}
	return r.insert(ctx, e)
}

// Close 排空队列并停掉后台 goroutine（测试收尾用）。可重复调用。
func (r *Recorder) Close() {
	if r == nil {
		return
	}
	r.mu.Lock()
	if r.closed {
		r.mu.Unlock()
		return
	}
	r.closed = true
	close(r.queue)
	r.mu.Unlock()
	r.wg.Wait()
}

// Dropped 返回因队列满而丢弃的行数（测试断言"满队列不阻塞"用）。
func (r *Recorder) Dropped() int64 {
	if r == nil {
		return 0
	}
	return r.dropped.Load()
}

// loop 是唯一的消费者：串行落库，失败只记日志——审计写失败不回滚业务写入（契约 §3）。
func (r *Recorder) loop() {
	defer r.wg.Done()
	for e := range r.queue {
		ctx, cancel := context.WithTimeout(context.Background(), insertTimeout)
		err := r.insert(ctx, e)
		cancel()
		if err != nil {
			slog.Error("审计落库失败：丢弃该行（业务写入不回滚）",
				"service", r.service, "action", e.Action, "request_id", e.RequestID, "err", err)
		}
	}
}

const insertSQL = "INSERT INTO audit.audit_log (id, occurred_at, service, action, actor_user_id, actor_username, credential_type, actor_ip, actor_user_agent, target_type, target_id, changes, result, error_code, request_method, route, http_status, request_id) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18)"

// insert 落一行：补齐默认值 → 脱敏 changes → 写库。四个入口（Record/RecordSync）都经这里，
// 保证"写入前必须过滤"（契约 §4）没有旁路。
func (r *Recorder) insert(ctx context.Context, e Entry) error {
	if r.db == nil {
		return errors.New("audit: recorder has no database")
	}
	if e.ID == "" {
		e.ID = uuid.NewString()
	}
	if e.OccurredAt.IsZero() {
		e.OccurredAt = time.Now().UTC()
	}
	if e.Service == "" {
		e.Service = r.service
	}
	if e.Result == "" {
		e.Result = "success"
	}
	payload, err := encodeChanges(SanitizeChanges(e.Changes))
	if err != nil {
		return err
	}
	// actor_user_id 是 uuid 列：为空写 NULL；不是 uuid 时保行丢 id（用户名快照仍在），
	// 否则整行会被 Postgres 拒掉——留痕宁可少一列，也不能整条丢。
	var actorID any
	if raw := strings.TrimSpace(e.ActorUserID); raw != "" {
		parsed, perr := uuid.Parse(raw)
		if perr != nil {
			slog.Warn("审计：actor_user_id 不是 uuid，按空值写入", "service", r.service, "action", e.Action, "actor_user_id", raw)
		} else {
			actorID = parsed.String()
		}
	}
	_, err = r.db.ExecContext(ctx, insertSQL,
		e.ID, e.OccurredAt, e.Service, e.Action, actorID, e.ActorUsername, e.CredentialType,
		e.ActorIP, truncate(e.ActorUserAgent, userAgentMax), e.TargetType, e.TargetID, payload,
		e.Result, e.ErrorCode, e.RequestMethod, e.Route, e.HTTPStatus, e.RequestID)
	return err
}

// Describe 由处理器调用：登记被动对象与"变更前后"摘要。
func Describe(c *gin.Context, d Detail) {
	c.Set(detailKey, d)
}

// Fail 由处理器/闸门调用：登记失败错误码，值必须与响应体的 error 字段一致；
// 中间件据此写 result=failure + error_code（没登记时回落 http_<status>）。
func Fail(c *gin.Context, code string) {
	if code != "" {
		c.Set(ErrorCodeKey, code)
	}
}

// Middleware 给写路由接线。挂载位置很关键：必须在身份中间件**之后**（草稿里的 actor 要在
// c.Next() 之前可读），且在写路由注册**之前**（gin 的 RouterGroup.Use 只对之后注册的路由生效）。
//
// 同一个引擎可以挂多个实例，各自的动作码表不重叠即可（没有别的办法覆盖注册在路由组之外的
// 端点：引擎级中间件在 c.Next() 之前拿不到组内中间件解析出的身份）。
func Middleware(o Options) gin.HandlerFunc {
	actions := o.Actions
	if actions == nil {
		actions = map[string]string{}
	}
	return func(c *gin.Context) {
		if o.Recorder == nil {
			// 未注入 recorder（单测拿不到库）时不写审计，也不改变请求语义。
			c.Next()
			return
		}
		// 路由模板而不是原始路径：原始路径没有额外信息，模板才能聚合（契约 §1）。
		// FullPath 为空说明没有匹配到路由（404 之类），此时也不该有动作码。
		route := c.FullPath()
		if route == "" {
			route = c.Request.URL.Path
		}
		action, ok := actions[c.Request.Method+" "+route]
		if !ok {
			c.Next()
			return
		}
		e := Entry{
			OccurredAt:     time.Now().UTC(),
			Action:         action,
			ActorIP:        c.ClientIP(),
			ActorUserAgent: c.GetHeader("User-Agent"),
			RequestMethod:  c.Request.Method,
			Route:          route,
			RequestID:      requestID(c),
		}
		if o.Actor != nil {
			a := o.Actor(c)
			e.ActorUserID, e.ActorUsername, e.CredentialType = a.UserID, a.Username, a.CredentialType
		}
		c.Next()
		status := c.Writer.Status()
		if status == 0 {
			status = http.StatusOK
		}
		e.HTTPStatus = status
		if status < 400 {
			e.Result = "success"
		} else {
			e.Result = "failure"
			e.ErrorCode = c.GetString(ErrorCodeKey)
			if e.ErrorCode == "" {
				e.ErrorCode = fmt.Sprintf("http_%d", status)
			}
		}
		if v, ok := c.Get(detailKey); ok {
			if d, ok := v.(Detail); ok {
				e.TargetType, e.TargetID, e.Changes = d.TargetType, d.TargetID, d.Changes
			}
		}
		o.Recorder.Record(e)
	}
}

// requestID 取 X-Request-Id 透传值，缺失时生成 uuid 并回写响应头（契约 §1）。
func requestID(c *gin.Context) string {
	rid := strings.TrimSpace(c.GetHeader(RequestIDHeader))
	if rid == "" {
		rid = uuid.NewString()
		c.Header(RequestIDHeader, rid)
	}
	return rid
}

File: frontend/src/lib/api/audit.ts
// 本人操作记录：读取面是账号服务的 GET /api/admin/audit-logs（跨服务审计表 audit.audit_log）。
//
// 同一个端点在服务端分两档（契约 docs/architecture/audit-log.md §5）：持 auth.audit.read 者按
// 全量过滤条件查（管理台排障），其余登录用户被强制收敛到 actor_user_id = 自己。本模块**不带**
// actor_user_id 参数——作用域由服务端按会话决定，前端不复制一份"我是不是管理员"的判定：那种判定
// 一旦与服务端不一致，要么变成越权读取，要么让本人读不到自己的记录。
//
// 跨用户过滤（actor_user_id 指向他人、或 actor 前缀）服务端回 403 而不是静默改写成自己，
// 调用方必须当失败处理，不能当成"查到了别人的记录"。
import { fetchApi } from "./client";
import { requireArray, safeCount } from "./fields";

/** 与 audit.audit_log 的列一一对应；changes 已由写入侧脱敏（"[redacted]" / "a***@domain"）。 */
export interface AuditLogEntry {
  id: string;
  occurred_at: string;
  /** 产生这条记录的服，取值 catalog / auth / community / storage。 */
  service: string;
  /** 稳定机器码，形如 <域>.<过去式动作>（如 entity.saved、session.login）。只增不改。 */
  action: string;
  actor_user_id?: string;
  actor_username?: string;
  /** session / pat / oauth / anonymous：区分"用登录会话做的"与"用令牌做的"。 */
  credential_type?: string;
  actor_ip?: string;
  actor_user_agent?: string;
  target_type?: string;
  target_id?: string;
  /** 变更摘要，已脱敏并截断；空对象表示这次操作没有可摘要的字段变化。 */
  changes?: Record<string, unknown>;
  /** success / failure。 */
  result?: string;
  /** 失败原因码（与响应体的 error 同值，如 forbidden、invalid_payload）。 */
  error_code?: string;
  request_method?: string;
  route?: string;
  http_status?: number;
  /** 与请求头 X-Request-Id 同值，排障时用它对齐日志。 */
  request_id?: string;
}

export interface AuditLogPage {
  items: AuditLogEntry[];
  /** 与服务端同一套过滤条件下的总行数，不是本页条数。 */
  total: number;
  page: number;
  per_page: number;
}

/** 与账号服务的响应形状一致：单页 20 条足够读完一屏，翻页由调用方控制。 */
export const AUDIT_PAGE_SIZE = 20;

export async function fetchOwnAuditLogs(page = 1, perPage = AUDIT_PAGE_SIZE): Promise<AuditLogPage> {
  const params = new URLSearchParams({ page: String(page), per_page: String(perPage) });
  const res = await fetchApi<{ items?: unknown[]; total?: number; page?: number; per_page?: number }>(
    "/admin/audit-logs?" + params.toString()
  );
  // 界面上有"暂无记录"文案，所以取不到 items 必须抛（requireArray），不能折成空数组——
  // 那会把"读不到"讲成"你没做过任何操作"（同 lib/api/fields.ts 的口径）。
  const rows = requireArray<unknown>(res?.items, "items");
  const currentPage = safeCount(res?.page, page);
  const currentSize = safeCount(res?.per_page, perPage);
  return {
    items: rows.map(normalizeAuditEntry),
    total: safeCount(res?.total, rows.length),
    // 服务端保证 page>=1 / per_page>=1；真回 0 时按请求值兜底，避免分页控件算出零页。
    page: currentPage < 1 ? page : currentPage,
    per_page: currentSize < 1 ? perPage : currentSize,
  };
}

// normalizeAuditEntry 收敛一条记录：只给高频字段稳定类型，changes 原样透传——
// 它的形状由写入侧与契约决定，界面按 JSON 展示，不做二次加工（更不"还原"脱敏值）。
function normalizeAuditEntry(raw: unknown): AuditLogEntry {
  const row = (raw ?? {}) as Record<string, any>;
  return {
    id: String(row.id ?? ""),
    occurred_at: typeof row.occurred_at === "string" ? row.occurred_at : "",
    service: str(row.service) ?? "",
    action: str(row.action) ?? "",
    actor_user_id: str(row.actor_user_id),
    actor_username: str(row.actor_username),
    credential_type: str(row.credential_type),
    actor_ip: str(row.actor_ip),
    actor_user_agent: str(row.actor_user_agent),
    target_type: str(row.target_type),
    target_id: str(row.target_id),
    changes:
      row.changes && typeof row.changes === "object" && !Array.isArray(row.changes)
        ? (row.changes as Record<string, unknown>)
        : undefined,
    result: str(row.result),
    error_code: str(row.error_code),
    request_method: str(row.request_method),
    route: str(row.route),
    http_status: typeof row.http_status === "number" ? row.http_status : undefined,
    request_id: str(row.request_id),
  };
}

function str(value: unknown): string | undefined {
  return typeof value === "string" && value !== "" ? value : undefined;
}
